language/evm/hardhat-move/src/index.ts

```diff
@@ -1,392 +1,99 @@
-    if (buildRes.isErr()) {
-        let e = buildRes.error;
-        console.log(`\nFailed to build ${packagePath}\n${e.stdout}${e.stderr}`);
-        return err(e);
-    }
-
-    let genArtifactsRes = await generateArtifactsForPackage(hardhatRootPath, packagePath);
-    if (genArtifactsRes.isErr()) {
-        let e = genArtifactsRes.error;
-        console.log(`Failed to build ${packagePath}\n${e}`);
-        return err(genArtifactsRes.error);
-    }
-
-    console.log(`Successfully built ${packagePath}`);
-
-    return ok(genArtifactsRes.value);
-}
-
-/***************************************************************************************
- *
- * Move Compile Subtask (Entrypoint)
- *
- *   This adds a new subtask "compile:move" which is added to the queue when one runs
- *   `npx hardhat compile`. This task will build all the move contracts using the `move`
- *   executable and generate the artifacts hardhat requires for testing and deployment.
- *
- **************************************************************************************/
-const TASK_COMPILE_MOVE: string = "compile:move";
-
-subtask(
-    TASK_COMPILE_GET_COMPILATION_TASKS,
+// This adds a new subtask "compile:move" which is added to the queue when one runs
+// `npx hardhat compile`. This task will build all the move contracts using the `move`
+// executable and generate the artifacts hardhat requires for testing and deployment.
+subtask(TASK_COMPILE_GET_COMPILATION_TASKS,
     async (_, __, runSuper): Promise<string[]> => {
         const otherTasks = await runSuper();
         return [...otherTasks, TASK_COMPILE_MOVE];
-    }
-);
+    });
 
 subtask(TASK_COMPILE_MOVE)
     .addParam("quiet", undefined, undefined, types.boolean)
     .setAction(async ({ quiet }: { quiet: boolean }, { artifacts, config, run }) => {
-        let packagePaths: String[] = await listMovePackages(Path.join(config.paths.root, "contracts"));
-
-        if (packagePaths.length == 0) {
-            console.log("No Move contracts to compile");
-            return;
+        const arch = config.move.arch;
+        const compilerPath = config.move.compilerPath;
+
+        let movePath: string;
+        if (compilerPath) {
+            movePath = compilerPath;
+            console.log("using local move executable:", movePath);
+        } else {
+            const moveBuild: MoveBuild = await run(TASK_COMPILE_MOVE_GET_BUILD);
+            movePath = moveBuild.compilerPath;
+            console.log("using auto find move executable:", movePath);
         }
 
-        let plural = packagePaths.length == 1 ? "" : "s";
-        console.log("Building %d Move package%s...", packagePaths.length, plural);
+        await run(
+            TASK_COMPILE_MOVE_RUN_BINARY,
+            {
+                arch: arch,
+                movePath: movePath,
+                artifacts: artifacts,
+                config: config
+            }
+        );
+    });
+
+subtask(TASK_COMPILE_MOVE_RUN_BINARY)
```

Comment:
I think it is only for the purpose of maintenance and reviewing, it will be more clear to split a "big" task into separate sub tasks from my perspective